### ISSUE TITLE
Generate Rust code representing GRIB2 tables from GDAL CSVs

### ISSUE THREAD

**Issue by JackKelly** [mpiannucci/gribberish#63]
In [this comment](https://github.com/mpiannucci/gribberish/issues/41#issuecomment-2405209996) @mpiannucci said:

> I... was going to codegen [a Rust implementation of the GRIB code tables] but never had the time

We could use this issue to capture some ideas & design requirements for the codegen. I'll get started below but please shout if I'm off track :slightly_smiling_face:.

## Input to the codegen
[`GDAL's CSV representation of the GRIB tables`](https://github.com/OSGeo/gdal/tree/master/frmts/grib/data). The README for that directory is here: [`gdal/frmts/grib/degrib/README.TXT`](https://github.com/OSGeo/gdal/blob/master/frmts/grib/degrib/README.TXT)

## Output from the codegen
Rust code!
- Rust Enums which represent the GRIB code tables.
- Make it super-easy and super-fast to look up the appropriate enum variant by generating the definitions for at least 2 [`phf::Map`](https://docs.rs/phf/latest/phf/)s (perfect hash functions generated at compile time):
    1. Map from abbreviation `String` to enum variant. This will be useful for parsing `.idx` files (#41).
    2. Map from numeric GRIB code to enum variant.

## Questions
1. How to handle the fact that GRIB code table 4.2 "Parameter number" allows each weather centre to use a certain range of codes for their own use? Some options:
    1. Generate one hashmap per weather centre.
        - Pros: Very fast at runtime & very easy for users.
        - Cons: Space-inefficient. The shared codes will be repeated _n_ times, where _n_ is the number of weather centres.
    2. Have one hashmap which captures all the common codes. And then have _n_ smaller hashmaps which capture the centre-specific codes.
        - Pros: Space-efficient
        - Cons: A little slower at runtime, and a little harder for users. e.g. to look up an abbreviation, the user will first search the common hashmap. If it returns `None` then look up in the centre-specific map.
        - (I think I'd lean towards this approach?)
2. Where should the codegen crate live? In `gribberish/codegen/`?
3. What's the key for the map that decodes the numerical GRIB code? Is it a `struct Key {prod: u8, cat: u8, subcat: u8}`?
4. To decode `.idx` files, we need a single hashmap which holds _every_ abbreviation string. So the `values` of the hashmap have to all be the same type. Do we represent the different GRIB product categories (e.g. "temperature" & "moisture") as different enums (which is how `gribberish` currently works) and then have a "super" enum which embeds all the product categories into a single "top level" enum like this:

```rust
enum Product {
    Temperature(Temperature),
    Moisture(Moisture),
    // etc...
}
```

So the "abbrev decoding" `phf::Map` would be:

```rust
static ABBREV_TO_PRODUCT: phf::Map<&'static str, Product> = phf::phf_map! {
    "TMP" => Temperature(Temperature::Temperature),
    "VTMP" => Temperature(Temperature::VirtualTemperature),
    // etc...
};
```
5. What should each enum variant look like? Would we continue using attribute macros like `#[name = "surface"]` and `#[description = "ground or water surface"]`? Or, once we're generating the enums, could we drop the macros and use something like this?

```rust
struct ProductVariant {
    prod: u8,
    cat: u8,
    subcat: u8,
    short_name: String,
    name: String,
    unit: Unit,
    unit_conv: UnitConv,
}

enum Temperature {
    Temperature(ProductVariant),
    VirtualTemperature(ProductVariant),
    // etc...
}
```

## Development timeline
Hopefully I'll be able to help write the PR which implements this in a few months... But, of course, I'm more than happy for other people to do it!

## Related
- ["Introduction to Code Generation in Rust"](https://crabnebula.dev/blog/introduction-to-code-generation-in-rust/) by Chip Reed at CrabNebula, November 2023. (Discusses very useful crates for us: `quote`, `walkdir`, and `phf`)

**Comment by JackKelly** [comment]
**UPDATE**: I've removed the questions for now because I'm working on a code sketch which I hope answers the main questions... I'll post a comment with the code sketch soon...

**Comment by JackKelly** [comment]
Huh. I've just stumbled across the [`grib-rs`](https://github.com/noritada/grib-rs) project (another GRIB reader implemented in Rust :slightly_smiling_face:). `grib-rs` includes some codegen. Some notable parts of `grib-rs`:

- [`gen/src/grib2_codeflag_csv.rs`](https://github.com/noritada/grib-rs/blob/master/gen/src/grib2_codeflag_csv.rs): Parse WMO CSV files into a vector of [`Wgrib2TableEntry`](https://github.com/noritada/grib-rs/blob/a9b73ce1ab2011e1899539d60baebd3117aa76ae/codegen/src/param_codes.rs#L93-L105) structs.
- [`codegen/src/param_codes.rs`](https://github.com/noritada/grib-rs/blob/master/codegen/src/param_codes.rs): Generate enum variants from a vector of [`Wgrib2TableEntry`](https://github.com/noritada/grib-rs/blob/a9b73ce1ab2011e1899539d60baebd3117aa76ae/codegen/src/param_codes.rs#L93-L105) structs. I _think_ the generated code is never materialised to disk. I think it's just generated on-the-fly during compilation.
- [`codegen/tests/01-normal-case.rs`](https://github.com/noritada/grib-rs/blob/a9b73ce1ab2011e1899539d60baebd3117aa76ae/codegen/tests/01-normal-case.rs): An example of a generated `enum`.

(note that, in the `Wgrib2TableEntry` struct, what `gribberish` would call the `abbrev` is called the `name`).

AFAICT, the representation of the GRIB tables in `grib-rs` is "flat":  `grib-rs` doesn't attempt to represent the hierarchy "Discipline > Category > Product > Parameter".

I don't _think_ `grib-rs` currently has any ability to read `.idx` files. So you'd have to use a `HashMap<String, Wgrib2TableEntry>` to find the parameter associated with each "abbreviation" string in an `.idx` file. Which might be fine, from a performance perspective.

I must admit I'm leaning towards the idea that it would be useful to have a common "grib_tables" crate which "just" represents the GRIB tables as Rust code (and the auto-generated code would be materialised to disk).

**Comment by mpiannucci** [comment]
Yeah i know about grib-rs. Its similar but there is no python support, and because of the way it defines code tables its limited to WMO tables which was not sufficient for me

**Comment by JackKelly** [comment]
>  I'll post a comment with the code sketch soon

I've been back-and-forth on this design.

TL;DR: I think I'm gonna implement a simple stand-alone crate that loads the WMO and GDAL CSVs at runtime and allows folks to look up parameters by their numeric ID or by their abbreviation. More details here: https://github.com/JackKelly/hypergrib/tree/main/crates/grib_tables

I had previously started work on [some code](https://github.com/JackKelly/rust-playground/tree/main/grib_tables) that tries to faithfully capture the GRIB tables hierarchy (Discipline > Category > Parameter). But, for `hypergrib`, this feels overkill. It might be slightly faster because we can create a perfect hash at compile time. But then I realised that we only need to look things up in the GRIB table whilst indexing a new GRIB dataset. When "normal users" use the dataset via `hypergrib`, they can just read the metadata that we create for the dataset.

**Comment by emfdavid** [comment]
Heads up [eccodes just changed](https://github.com/ecmwf/eccodes/commit/9f0432deb0d0f57c8ac002b2d14236980d444674#diff-bf52f84770b011c828396c842f5a282899bede6bb3acd2c967b50eabca327e42L164-R165) the variable names for NOAA grib2 files.
dswrf -> sdswrf
ulwrf -> sulwrf
etc...
Still trying to work out if this is intentional or a bug.

**Comment by williamhobbs** [comment]
@emfdavid, did you figure out if those variable name changes in eccodes were intentional?
